chain/chain/src/store.rs

```diff
@@ -480,35 +482,101 @@ impl ChainStore {
-                } else {
-                    shard_id
-                };
-                let mut receipts = self
-                    .get_outgoing_receipts(&receipts_block_hash, receipts_shard_id)
-                    .map(|v| v.to_vec())
-                    .unwrap_or_default();
-
-                // filter to receipts that belong to `shard_id` in the current shard layout
-                if shard_layout != receipts_shard_layout {
-                    receipts.retain(|receipt| {
-                        account_id_to_shard_id(&receipt.receiver_id, &shard_layout) == shard_id
-                    });
-                }
+            if block_header.height() != last_included_height {
+                receipts_block_hash = *block_header.prev_hash();
+                continue;
+            }
+            let receipts_shard_layout = epoch_manager.get_shard_layout(block_header.epoch_id())?;
 
-                return Ok(receipts);
+            // get the shard from which the outgoing receipt were generated
+            let receipts_shard_id = if shard_layout != receipts_shard_layout {
+                shard_layout.get_parent_shard_id(shard_id)?
             } else {
-                receipts_block_hash = *block_header.prev_hash();
+                shard_id
+            };
+
+            let mut receipts = self
+                .get_outgoing_receipts(&receipts_block_hash, receipts_shard_id)
+                .map(|v| v.to_vec())
+                .unwrap_or_default();
+
+            if shard_layout != receipts_shard_layout {
+                #[cfg(feature = "protocol_feature_simple_nightshade_v2")]
+                Self::reassign_outgoing_receipts_for_resharding_v2(
+                    &shard_layout,
+                    shard_id,
+                    receipts_shard_id,
+                    &mut receipts,
+                );
+
+                #[cfg(not(feature = "protocol_feature_simple_nightshade_v2"))]
+                Self::reassign_outgoing_receipts_for_resharding_v1(
+                    &shard_layout,
+                    shard_id,
+                    &mut receipts,
+                );
             }
+
+            return Ok(receipts);
         }
     }
 
+    /// Reassign the outgoing receipts from the parent shard to the children
+    /// shards.
+    ///
+    /// This method does it based on the "lowest child index" approach where it
+    /// assigns all the receipts from parent to the child shard with the lowest
+    /// index. It's meant to be used for the resharding from simple nightshade
+    /// with 4 shards to simple nightshade v2 with 5 shards and subsequent
+    /// reshardings.
+    #[cfg(feature = "protocol_feature_simple_nightshade_v2")]
+    fn reassign_outgoing_receipts_for_resharding_v2(
+        shard_layout: &ShardLayout,
+        shard_id: ShardId,
+        parent_shard_id: ShardId,
+        receipts: &mut Vec<Receipt>,
+    ) {
+        let split_shard_ids = shard_layout.get_split_shard_ids(parent_shard_id);
```

Comment:
why not use an expect/unwrap here?

I suppose it's fine given we are only calling this when the receipt shard layout is different from the current shard layout.

Reply:
I'm handling errors in the usual way, by returning result. Just to be sure I will check what happens if that particular error happens. Generally speaking unwraps and expects are not great because there is no way to recover and it takes down the binary immediately.